pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/RecoverableMultithreadedProcessor.java

```diff
@@ -0,0 +1,279 @@
+            readerGroupManager = ReaderGroupManager.withScope(scope, controllerURI);
+
+            ReaderGroupConfig.ReaderGroupConfigBuilder builder = ReaderGroupConfig.builder()
+                    .disableAutomaticCheckpoints();
+
+            // Attempt to load the last Pravega checkpoint.
+            try {
+                String checkpointName = StandardCharsets.UTF_8.decode(ByteBuffer.wrap(Files.readAllBytes(latestCheckpointNamePath))).toString();
+                Path checkpointDirPath = checkpointRootPath.resolve(checkpointName);
+                Path checkpointPath = checkpointDirPath.resolve(PRAVEGA_CHECKPOINT_FILE_NAME);
+                log.info("Reading Pravega checkpoint from {}", checkpointPath);
+                Checkpoint checkpoint = Checkpoint.fromBytes(ByteBuffer.wrap(Files.readAllBytes(checkpointPath)));
+                log.info("Starting from checkpointName={}, positions={}", checkpointName, checkpoint.asImpl().getPositions());
+                builder = builder.startFromCheckpoint(checkpoint);
+            } catch (Exception e) {
+                log.warn("Unable to load checkpoint from {}. Starting from the earliest event.", checkpointRootPath, e);
+                // This will create a reader group that starts from the earliest event.
+                builder = builder.stream(Stream.of(scope, inputStreamName));
+            }
+
+            final ReaderGroupConfig readerGroupConfig = builder.build();
+            readerGroupName = UUID.randomUUID().toString().replace("-", "");
+            readerGroupManager.createReaderGroup(readerGroupName, readerGroupConfig);
+            readerGroup = readerGroupManager.getReaderGroup(readerGroupName);
+
+            initiateCheckpointExecutor = Executors.newScheduledThreadPool(1);
+            performCheckpointExecutor = Executors.newScheduledThreadPool(1);
+            workerExecutor = Executors.newFixedThreadPool(numWorkers);
+
+            // Schedule periodic task to initiate checkpoints.
+            // If any execution of this task takes longer than its period, then subsequent executions may start late, but will not concurrently execute.
+            initiateCheckpointExecutor.scheduleAtFixedRate(this::performCheckpoint, checkpointPeriodMs, checkpointPeriodMs, TimeUnit.MILLISECONDS);
+        }
+
+        public void run() {
+            IntStream.range(0, numWorkers).forEach(workerIndex -> {
+                Worker worker = new Worker(workerIndex, scope, readerGroupName, outputStreamName, controllerURI);
+                workerExecutor.submit(worker);
+            });
+            try {
+                workerExecutor.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
+            } catch (InterruptedException e) {
+            }
+        }
+        /**
+         * Perform a checkpoint, wait for it to complete, and write the checkpoint to the state.
+         */
+        private void performCheckpoint() {
```

Comment:
Similar to the other class, this isn't really providing any value over automatic checkpoints. All this additional code defiantly makes the example more complex. It may be worth having one example of manual checkpoints somewhere, but it should be called out as one, and it should show a context where it is apparent, how they can provide value beyond what the automatic ones do.

Reply:
I will delete this entire file. However this comment applies equally to ExactlyOnceMultithreadedProcessor and I replied to it above.